### PR TITLE
Enhancing SFT Training Efficiency Using Packing and FlashAttention2 with Position IDs

### PR DESCRIPTION
**Pull request by RhuiDih**
# What does this PR do?

Improve throughput as well as train time and memory utilization for instruction tuning by enabling padding-free and attention mask-free attention. 

Specifically, this PR adds the capability to utilize `position_ids` in FlashAttention2 `_flash_attention_forward()` in the case of packing (`attention_mask=None`) to models which use `position_ids` in their respective `DecoderLayer` implementations.  

This PR also adds a new off-the-shelf data collator `DataCollatorWithFlattening` which packs the examples in a mini batch into one long sequence and return `position_ids` as well and turns the first token of `labels` to -100 to prevent the last token of previous example predicting the first token of next example.

This enables the following:
1. Use of packing for instruction tuning without incorrect cross-example attention 
2. Significant increase in training throughput  and reduction in memory utilization

**Example Result 1**
dataset: OrcaMath subset
setup: FSDP with 8 GPUs
| Model | DataProcess | Time | Throughput (token/s) | Memory (MB) |
| - | - | - | - | - |
| Llama2-7B | Padding  | 790 | 1269 | 22305 |
| Llama2-7B | ThisPR | 574 | 1746 | 20950 |
| Mistral-7B | Padding | 812 | 1216 | 23603 |
| Mistral-7B | ThisPR | 596 | 1658 | 22409 |

**Example Result 2**
dataset: FLAN subset
setup: FSDP with 8 GPUs
| Model | DataProcess | Time | Throughput (token/s) | Memory (MB) |
| - | - | - | - | - |
| Llama2-7B | Padding  | 1526 | 771 | 29234 |
| Llama2-7B | ThisPR | 809 | 1455 | 23854 |
| Mistral-7B | Padding | 742 | 742 | 30625 |
| Mistral-7B | ThisPR | 1408 | 1408 | 24549 |

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
Models:
- text models: @ArthurZucker

Edit - add more description on data collator

### CODE REVIEW

File: src/transformers/models/dbrx/modeling_dbrx.py [hunk 1]
```diff
@@ -495,9 +499,33 @@ def _flash_attention_forward(
 
             attn_output = pad_input(attn_output_unpad, indices_q, batch_size, query_length)
         else:
-            attn_output = flash_attn_func(
-                query_states, key_states, value_states, dropout, softmax_scale=softmax_scale, causal=causal
-            )
+            if (position_ids[:,-1]==position_ids.size(1)-1).all():
```
Comment (ArthurZucker) [inline src/transformers/models/dbrx/modeling_dbrx.py:502]: this is an input dependent control flow which won't be supported by compile, but compile is already not supported. 
This needs a comment: we are checking that the input is not padded, and that we are doing prefill right?

File: src/transformers/models/gemma/modeling_gemma.py [hunk 2]
```diff
@@ -49,6 +49,7 @@
 )
 from .configuration_gemma import GemmaConfig
 
+from ..llama.modeling_llama import prepare_fa2_from_position_ids
```
Comment (ArthurZucker) [inline src/transformers/models/gemma/modeling_gemma.py:52]: we never import from another modeling code in transformers, we use "copied from" or we define the general function in `falsh_attention_utils` for example

File: src/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -59,6 +59,20 @@
 _CONFIG_FOR_DOC = "LlamaConfig"
 
 
+def prepare_fa2_from_position_ids(query, key, value, position_ids, query_length):
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:62]: this needs to be documented to explain what is happening

**Review by ArthurZucker — COMMENTED** [review]
Thanks! I think we can wait a tad bit, #31446 is almost ready for merge!

**Comment by RhuiDih** [comment]
@ArthurZucker #31446 has just merged, let me rebase and update this PR

**Comment by RhuiDih** [comment]
@ArthurZucker Rebase is done.

I have move the codes to `modeling_flash_attention_utils.py` and added required comments/description.

Regarding the control flow, ideally the `cu_seq_len` should be prepared in data collator and model should accept it as arg to trigger `flash_attn_varlen_func` without relying on `position_ids`, but this is the quick & fast way to do padding free.

File: src/transformers/data/data_collator.py [hunk 4]
```diff
@@ -1611,3 +1611,29 @@ def numpy_mask_tokens(self, inputs: Any) -> Tuple[Any, Any, Any, Any]:
...
+
+@dataclass
+class DataCollatorWithFlattening(DefaultDataCollator):
+    return_tensors: str = "pt"
+    def __init__(self, return_position_ids=True):
+        self.return_position_ids=return_position_ids
+        warnings.warn(
+            "Using `DataCollatorForBatchFlattening` will flatten the entire mini batch into single long sequence."
+            "Make sure your attention computation is able to handles it!"
+        )
```
Comment (ArthurZucker) [inline src/transformers/data/data_collator.py:1623]: I think we keep track of a list of models that support FA2, if needed for guidance here!

File: src/transformers/modeling_flash_attention_utils.py [hunk 5]
```diff
@@ -203,6 +258,33 @@ def _flash_attention_forward(
             **flash_kwargs,
         )
         attn_output = pad_input(attn_output_unpad, indices_q, batch_size, query_length)
+
+    # if position_ids is provided and check not all examples (row) contain only 1 sequence,
+    # then use `flash_attn_varlen_func` to prevent cross-example attention and also allow padding free approach
+    elif position_ids is not None and not (position_ids[:,-1]==position_ids.size(1)-1).all():
```
Comment (ArthurZucker) [inline src/transformers/modeling_flash_attention_utils.py:264]: this is an input dependent control flow, but we already don't support compile with FA2 AFAIK. 
Alright for now!

**Review by ArthurZucker — COMMENTED** [review]
This LGTM! Only thing missing is a test! 
Testing this codepath in particuler, + datacollator in the trainer tests maybe?

**Comment by mayank31398** [comment]
I think loss computation is wrong for everything
it needs to be:
https://github.com/ibm-granite/dolomite-engine/blob/7e2527528ac54cb3388040a3052ef3b6d845a40d/dolomite_engine/hf_models/models/gpt_dolomite/main.py#L185-L191

**Comment by RhuiDih** [comment]
@mayank31398 
it can be done within the model or when labels preparation in data collator
https://github.com/huggingface/transformers/blob/808fd638182289cf1903a1d4c536f00c48809fb8/src/transformers/data/data_collator.py#L1634-L1637

**Comment by wynterl** [comment]
We are providing the capability to define the labels as the user wishes in the data collator in this PR via [transformers/src/transformers/data/data_collator.py](https://github.com/huggingface/transformers/blob/808fd638182289cf1903a1d4c536f00c48809fb8/src/transformers/data/data_collator.py#L1634-L1637)

**Comment by wynterl** [comment]
Hello @mayank31398 We are providing the capability to define the labels as the user wishes in the data collator in this PR via [transformers/src/transformers/data/data_collator.py](https://github.com/huggingface/transformers/blob/808fd638182289cf1903a1d4c536f00c48809fb8/src/transformers/data/data_collator.py#L1634-L1637)

**Comment by RhuiDih** [comment]
@ArthurZucker
I have added tests for the new data collator and the models which already have FA2 tests in place. For the trainer, i can't find existing FA2 + Trainer testing combo, could you point me there ?

**Review by ArthurZucker — APPROVED** [review]
LGTM, I just want a second look from our @fxmarty . 
This should improve performances for everyone. but let's make sure we don't break BC! WDYT @fxmarty ?

**Comment by ArthurZucker** [comment]
No worries I think the tests you added are enough

File: src/transformers/data/data_collator.py [hunk 6]
```diff
@@ -1611,3 +1611,30 @@ def numpy_mask_tokens(self, inputs: Any) -> Tuple[Any, Any, Any, Any]:
...
+@dataclass
+class DataCollatorWithFlattening(DefaultDataCollator):
+    return_tensors: str = "pt"
+    def __init__(self, *args, return_position_ids=True, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.return_position_ids=return_position_ids
+        warnings.warn(
+            "Using `DataCollatorWithFlattening` will flatten the entire mini batch into single long sequence."
+            "Make sure your attention computation is able to handle it!"
+        )
```
Comment (fxmarty) [inline src/transformers/data/data_collator.py:1624]: How should a user make sure?

**Review by fxmarty — COMMENTED** [review]
Looks good!

I strongly think this should be documented somewhere, maybe https://huggingface.co/docs/transformers/main/en/perf_train_gpu_one#flash-attention-2.

File: src/transformers/data/data_collator.py (hunk 6 above)
Comment (RhuiDih) [reply src/transformers/data/data_collator.py:1624]: no easy way, the user will have to make sure the model supports FA2 and have `position_ids`
until the model accepting `cu_seq_len` which will allow every model remove padding once for all, i think this serve as a good warning

File: tests/models/gemma/test_modeling_gemma.py [hunk 7]
```diff
@@ -526,6 +526,12 @@ def test_flash_attn_2_equivalence(self):
                 assert torch.allclose(logits_fa, logits, atol=3e-3)
 
 
+    @require_flash_attn
+    @require_torch_gpu
+    @slow
+    def test_flash_attention_2_padding_matches_padding_free_with_position_ids(self):
+        super().test_flash_attention_2_padding_matches_padding_free_with_position_ids()
+
```
[the remark this answers is not in the document]
Comment (RhuiDih) [reply tests/models/gemma/test_modeling_gemma.py]: i have removed them in latest commit

**Review by ArthurZucker — APPROVED** [review]
🚀 nice getting this done!

**Comment by RhuiDih** [comment]
@olisicky Hi, could you provide a minimal code to reproduce the error ? that would help greatly

**Comment by olisicky** [comment]
> @olisicky Hi, could you provide a minimal code to reproduce the error ? that would help greatly

Hi @RhuiDih . I found a mistake in my data as I was preparing a minimal code for you:D. So thank you for encouraging me to do so! As I am preparing data without `DataCollatorWithFlattening` I kept padding there to `context_length` even for bs=1 and it resulted in plenty of 0 in `position_ids` tensor on indexes corresponding to PAD tokens. 

However, I have an additional question. You are setting `-100` in labels for the first token in each sequence in the batch. I wonder why. I thought that the use of `position_ids` itself should reduce cross-example attention but here you mentioned that the -100 should prevent the last token of previous example predicting the first token of next example.

If I would have a sequences which start with BOS token and then it will be ignored during training by setting -100 in labels, then I would lost the importance of this first token. Is it necessary? Or maybe shouldn't we use some dummy token there rather then a part of the sequence itself?

Thank you very much!

**Comment by RhuiDih** [comment]
@olisicky 
The setting of `-100` is due to [causal loss implementation](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L1221-L1229), if we were flattening all sequences into batch of `1` without setting `-100` accordingly, it would incur unwanted loss.

First token of `input_ids` is still used to produce `logits`, hence no importance is lost. What we want to ignore, its the first token of `labels` due to the mentioned causal loss implementation.
Hope it is clear.

**Comment by olisicky** [comment]
> @olisicky The setting of `-100` is due to [causal loss implementation](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L1221-L1229), if we were flattening all sequences into batch of `1` without setting `-100` accordingly, it would incur unwanted loss.
> 
> First token of `input_ids` is still used to produce `logits`, hence no importance is lost. What we want to ignore, its the first token of `labels` due to the mentioned causal loss implementation. Hope it is clear.

Sorry for late reply. Yes, thank you. I will try it.

**Comment by MahmoudAshraf97** [comment]
Is packing usable in transformers `generate` method? AFAIK only padding is supported

**Comment by ArthurZucker** [comment]
Packing is supported with the collator introduced. We don't pack when generating tho, PRs are welcome for this!@

**Comment by qibao77** [comment]
> # What does this PR do?
> Improve throughput as well as train time and memory utilization for instruction tuning by enabling padding-free and attention mask-free attention.
> 
> Specifically, this PR adds the capability to utilize `position_ids` in FlashAttention2 `_flash_attention_forward()` in the case of packing (`attention_mask=None`) to models which use `position_ids` in their respective `DecoderLayer` implementations.
> 
> This PR also adds a new off-the-shelf data collator `DataCollatorWithFlattening` which packs the examples in a mini batch into one long sequence and return `position_ids` as well and turns the first token of `labels` to -100 to prevent the last token of previous example predicting the first token of next example.
> 
> This enables the following:
> 
> 1. Use of packing for instruction tuning without incorrect cross-example attention
> 2. Significant increase in training throughput  and reduction in memory utilization
> 
> **Example Result 1** dataset: OrcaMath subset setup: FSDP with 8 GPUs
> 
> Model	DataProcess	Time	Throughput (token/s)	Memory (MB)
> Llama2-7B	Padding	790	1269	22305
> Llama2-7B	ThisPR	574	1746	20950
> Mistral-7B	Padding	812	1216	23603
> Mistral-7B	ThisPR	596	1658	22409
> **Example Result 2** dataset: FLAN subset setup: FSDP with 8 GPUs
> 
> Model	DataProcess	Time	Throughput (token/s)	Memory (MB)
> Llama2-7B	Padding	1526	771	29234
> Llama2-7B	ThisPR	809	1455	23854
> Mistral-7B	Padding	742	742	30625
> Mistral-7B	ThisPR	1408	1408	24549
> ## Before submitting
> * [ ]  This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
> * [ ]  Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
>   Pull Request section?
> * [ ]  Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
>   to it if that's the case.
> * [ ]  Did you make sure to update the documentation with your changes? Here are the
>   [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
>   [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
> * [ ]  Did you write any new necessary tests?
> 
> ## Who can review?
> Models:
> 
> * text models: @ArthurZucker
> 
> Edit - add more description on data collator

How to use this feature for pretraining?

**Comment by RhuiDih** [comment]
@qibao7 usually one would truncate and pack to max length during pretraining, this PR will not benefit pretraining, unless most of your pretraining data are short
